curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -535,94 +538,130 @@ bool FileCacheManager::ReadKVRequestFromS3(const std::string &name,
-            } else {
-                LOG(ERROR) << "read " << name << " fail";
-                return ret;
+                break;
             }
 
-            // update param
-            {
-                length -= currentReadLen;         // Remaining read data length
-                readBufOffset += currentReadLen;  // next read offset
-                blockIndex++;
-                blockPos = (blockPos + currentReadLen) % blockSize;
-                objectOffset = 0;
+            if (ReadKVRequestFromS3(name, currentBuf, blockPos - objectOffset,
+                                    currentReadLen, &readRet)) {
+                VLOG(9) << "read " << name << " from s3 ok";
+                break;
             }
-        }
 
-        // add data to memory read cache
-        if (!curvefs::client::common::FLAGS_enableCto) {
-            auto chunkCacheManager = FindOrCreateChunkCacheManager(chunkIndex);
-            WriteLockGuard writeLockGuard(chunkCacheManager->rwLockChunk_);
-            DataCachePtr dataCache = std::make_shared<DataCache>(
-                s3ClientAdaptor_, chunkCacheManager, chunkPos, req->len,
-                dataBuf + req->readOffset, kvClientManager_);
-            chunkCacheManager->AddReadDataCache(dataCache);
+            LOG(ERROR) << "read " << name << " fail";
+            // make sure variable is set only once
+            std::call_once(cancelFlag, [&]() {
+                isCanceled.store(true);
+                ret.store(readRet);
+            });
+            return;
+        } while (false);
+
+        // update param
+        {
+            length -= currentReadLen;         // Remaining read data length
+            readBufOffset += currentReadLen;  // next read offset
+            blockIndex++;
+            blockPos = (blockPos + currentReadLen) % blockSize;
+            objectOffset = 0;
         }
     }
 
-    return 0;
+    // add data to memory read cache
+    if (!curvefs::client::common::FLAGS_enableCto) {
+        auto chunkCacheManager = FindOrCreateChunkCacheManager(chunkIndex);
+        WriteLockGuard writeLockGuard(chunkCacheManager->rwLockChunk_);
+        DataCachePtr dataCache = std::make_shared<DataCache>(
+            s3ClientAdaptor_, chunkCacheManager, chunkPos, req.len,
+            dataBuf + req.readOffset, kvClientManager_);
+        chunkCacheManager->AddReadDataCache(dataCache);
+    }
+}
+
+int FileCacheManager::ReadKVRequest(
+    const std::vector<S3ReadRequest> &kvRequests, char *dataBuf,
+    uint64_t fileLen) {
+    absl::BlockingCounter counter(kvRequests.size());
+    std::once_flag cancelFlag;
+    std::atomic_bool isCanceled{false};
+    std::atomic_int ret{0};
+
+    for (const auto &req : kvRequests) {
+        readTaskPool_.Enqueue([&]() {
+            auto defer = absl::MakeCleanup([&]() { counter.DecrementCount(); });
+            if (isCanceled) {
+                LOG(WARNING) << "kv request is canceled " << req.DebugString();
+                return;
```

Comment:
we need retry in `FileCacheManager::Read` , so if its failed here, we need return a meaningful value